### ISSUE TITLE
Make `person` and `person_distinct_id` deletion efficient

### ISSUE THREAD

**Issue by fuziontech** [PostHog/posthog#4242]
Mutations on our current replicated replacing merge tree have brought us down a few times now. It would be better to invest some time to finally migrate these tables over to be a replicated collapsing merge tree. 

Will fill in more details later. Going play with prototyping this now.

**Comment by macobo** [comment]
Related request: https://github.com/PostHog/posthog/issues/4260

**Comment by macobo** [comment]
We ended up going with a different strategy based on https://github.com/PostHog/product-internal/issues/47. Read more on why not the original approach above.

The end goal is to solve:
- problems with consistency on persons/person distinct_ids
- problems with performance (occasionally) caused by mutations on these two tables

The way we're going about this is:
- [x] Adding a new `is_deleted` column to persons in clickhouse https://github.com/PostHog/posthog/pull/4391
- [x] Update reads to respect `is_deleted` column
- [x] Update writes to populate `is_deleted` column https://github.com/PostHog/plugin-server/pull/393
- [ ] Do the same for person_distinct_id table as things have slowed down (once here, create a new issue)

**Comment by macobo** [comment]
Note: Also when organization gets deleted don't delete clickhouse events (but only persons, etc)

**Comment by macobo** [comment]
I have been looking into solving this, but am blocked. Problem is that `person_distinct_id` table is not what you think.

Most of our queries assume it's a mapping table from distinct_id <-> person_id, with distinct_id being unique However this invariant is not held in production, we use the following query in places (but not everywhere):

```
GET_LATEST_PERSON_DISTINCT_ID_SQL = """
SELECT * FROM person_distinct_id JOIN (
    SELECT distinct_id, max(_offset) as _offset FROM person_distinct_id WHERE team_id = %(team_id)s GROUP BY distinct_id
) as person_max ON person_distinct_id.distinct_id = person_max.distinct_id AND person_distinct_id._offset = person_max._offset
WHERE team_id = %(team_id)s
"""
```

This indicates a bug in the ingestion pipeline, which needs to be solved before we can move to collapsingmergetree on the table.

**Comment by macobo** [comment]
Digging into this, discovered a few things:

1. The data issue will be fixed by solving this issue: https://github.com/PostHog/product-internal/issues/2
2. There's a minor race condition in plugin-server which affects this: https://github.com/PostHog/plugin-server/issues/502

As-is, I believe I have all the context needed to solve and ship this. :) The only hurdle should be updating all queries inside posthog/posthog.
